src/server/pps/server/s3g_sidecar.go

```diff
@@ -0,0 +1,475 @@
+				},
+				Spec: v1.ServiceSpec{
+					Selector: labels,
+					// Create a headless service so that the worker's kube proxy doesn't
+					// have to get a routing path for the service IP (i.e. the worker kube
+					// proxy can have stale routes and clients running inside the worker
+					// can still connect)
+					ClusterIP: "None",
+					Ports: []v1.ServicePort{
+						{
+							Port: int32(s.apiServer.env.S3GatewayPort),
+							Name: "s3-gateway-port",
+						},
+					},
+				},
+			}
+
+			err := backoff.RetryNotify(func() error {
+				_, err := s.apiServer.env.GetKubeClient().CoreV1().Services(s.apiServer.namespace).Create(service)
+				if err != nil && strings.Contains(err.Error(), "already exists") {
+					return nil // service already created
+				}
+				return err
+			}, backoff.NewExponentialBackOff(), func(err error, d time.Duration) error {
+				logrus.Errorf("error creating kubernetes service for s3 gateway sidecar: %v; retrying in %v", err, d)
+				return nil
+			})
+			if err != nil {
+				logrus.Errorf("could not create service for %q: %v", jobInfo.Job.ID, err)
+			}
+		},
+
+		OnTerminate: func(_ context.Context, jobID string) {
+			if !ppsutil.ContainsS3Inputs(s.pipelineInfo.Input) && !s.pipelineInfo.S3Out {
+				return // Nothing to delete; this isn't an s3 pipeline (shouldn't happen)
+			}
+
+			if err := backoff.RetryNotify(func() error {
+				err := s.apiServer.env.GetKubeClient().CoreV1().Services(s.apiServer.namespace).Delete(
+					ppsutil.SidecarS3GatewayService(jobID),
+					&metav1.DeleteOptions{OrphanDependents: new(bool) /* false */})
+				if err != nil && strings.Contains(err.Error(), "not found") {
+					return nil // service already deleted
+				}
+				return err
+			}, backoff.NewExponentialBackOff(), func(err error, d time.Duration) error {
+				logrus.Errorf("error deleting kubernetes service for s3 %q gateway sidecar: %v; retrying in %v", jobID, err, d)
+				return nil
+			}); err != nil {
+				logrus.Errorf("permanent error deleting kubernetes service for %q s3 gateway sidecar: %v", jobID, err)
+			}
+		},
+	}
+}
+
+func (h *jobHandler) start() {
+	defer func() {
+		logrus.Errorf("sidecar s3 gateway: start() is exiting; this should not happen")
+		panic(
+			fmt.Sprintf("!sidecar s3 gateway start() is exiting; this should never happen"),
+		)
+	}()
+establish_watch:
+	for { // reestablish watch in a loop, in case there's a watch error
+		var watcher watch.Watcher
+		backoff.Retry(func() error {
+			var err error
+			watcher, err = h.s3g.apiServer.jobs.ReadOnly(context.Background()).WatchByIndex(
+				ppsdb.JobsPipelineIndex, h.s3g.pipelineInfo.Pipeline)
+			if err != nil {
+				return fmt.Errorf("error creating watch: %v", err)
+			}
+			return nil
+		}, backoff.NewInfiniteBackOff())
+		defer watcher.Close()
+
+		for e := range watcher.Watch() {
+			jobID := string(e.Key)
+			if e.Type == watch.EventError {
+				logrus.Errorf("sidecar s3 gateway watch error: %v", e.Err)
+				continue establish_watch
```

Comment:
I found this structure easy to follow visually when I was writing it (because if the `if` inside the `for`) but I really don't feel strongly about it so I changed it.